File: sort/quick_sort.cpp
// +build ignore
// from golang

#include "quick_sort.h"

// quick sort - asc
void quickSort(int a[], int low, int high) {
    if(low >= high) return;
    int first = low, last = high, key = a[first];
    while(first < last) {
        while(first < last && a[last] >= key) --last;
        a[first] = a[last];
        while(first < last && a[first] <= key) ++first;
        a[last] = a[first];    
    }
    a[first] = key;
    quickSort(a, low, first-1);
    quickSort(a, first+1, high);
}

File: sort/quick_sort.go
package sort

// quick sort - asc
func QuickSort(inArr []int) {
	if len(inArr) <= 1 {
		return
	}
	mid, i, head, tail := inArr[0], 1, 0, len(inArr)-1
	for head < tail {
		if inArr[i] > mid {
			inArr[i], inArr[tail] = inArr[tail], inArr[i]
			tail--
			continue
		}
		inArr[i], inArr[head] = inArr[head], inArr[i]
		head++
		i++
	}
	inArr[head] = mid
	QuickSort(inArr[:head])
	QuickSort(inArr[head+1:])
}
